src/sqlfluff/core/parser/segments/base.py

```diff
@@ -632,19 +665,24 @@ def is_optional(cls):
     @classmethod
     def class_is_type(cls, *seg_type):
         """Is this segment class (or its parent) of the given type."""
-        # Do we match on the type of _this_ class.
-        if cls.type in seg_type:
-            return True
-        # If not, check types of parents.
-        for base_class in cls.__bases__:
-            if base_class is object:
-                break
-            elif base_class.type in seg_type:
+        # Work backward in the class heirachy to see if there's a match
+        for class_type in cls._class_types():
```

Comment:
```suggestion
        # Work backward in the class hierarchy to see if there's a match
        for class_type in cls._class_types():
```

Note if we precompute `_class_types()`, this loop can be replaced with a simple set-based check.